spring-data-jdbc/src/main/java/org/springframework/data/jdbc/core/mapping/IdGeneratingBeforeSaveCallback.java

```diff
@@ -45,26 +45,47 @@ public Object onBeforeSave(Object aggregate, MutableAggregateChange<Object> aggr
         Assert.notNull(aggregate, "The aggregate cannot be null at this point");
 
         RelationalPersistentEntity<?> persistentEntity = relationalMappingContext.getPersistentEntity(aggregate.getClass());
+
+        if (persistentEntity.getIdProperty() == null) {
+            return aggregate;
+        }
+
+        // we're doing INSERT and ID property value is not set explicitly by client
+        if (persistentEntity.isNew(aggregate) && !idIsProvidedByClient(aggregate, persistentEntity)) {
+            return potentiallyFetchIdFromSequence(aggregate, persistentEntity);
+        } else {
+            return aggregate;
+        }
+    }
+
+    private boolean idIsProvidedByClient(Object aggregate, RelationalPersistentEntity<?> persistentEntity) {
+        Object identifier = persistentEntity.getIdentifierAccessor(aggregate).getIdentifier();
+
+        if (persistentEntity.getIdProperty().getType().isPrimitive()) {
+            return identifier instanceof Number num && num.longValue() != 0L;
```

Comment:
Notice to myself: We should introduce an utility to determine whether a primitive is set to its initial value.

Reply:
I can introduce it **in separate commit** in the same PR, it seems that it is natural to do it in the context of the current ticket. What do you think @mp911de ?

Reply:
I was considering an utility in Spring Data Commons. Let me verify whether this is a good idea or not.

Reply:
Sure, I'll wait for your response. I think I can help with this as well.